bayes_kit/rhat.py

```diff
@@ -1,11 +1,88 @@
 import numpy as np
 from numpy.typing import NDArray, ArrayLike
+import scipy as sp
 
 FloatType = np.float64
 VectorType = NDArray[FloatType]
 SeqType = ArrayLike
 
 
+def split_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
+    are ascending and start with 1 for the smallest value.
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
+    reshaped_arrays = []
+    current_index = 0
+    for array in chains:
+        size = len(array)
```

Comment:
ArrayLike contains all sorts of things you can’t call len() on, like `int`. For most of our purposes we probably shouldn’t be using it, though these won’t really be caught until #26

Reply:
Any suggestion on what I should be using for typing instead?

Reply:
It's a judgement call, but I'm in favor of using our own types instead of `ArrayLike` for things like this. `Union[typing.Sequence[float], npt.NDArray[np.float64]]` is probably the minimal type we actually will pass for a chain.

Writing out our own type has some downsides (like we might need to broaden it over time if we also wanted to pass e.g. `list[int]`) but is useful if only because we can avoid things being in it that we don't want.

If we want to stick with `npt.ArrayLike`, you essentially just need to insert `np.asarray` or `np.asanyarray` calls everywhere before doing anything, since `ArrayLike` is really only defined as "things you can pass to `np.asarray`"

Reply:
I went with @WardBrian's suggestion for now. We may eventually want to go with something that'd work with single precision, but then we need to be careful throughout not to upcast.

@WardBrian also pointed out that the function will still work with `np.float32` input, it just won't pass mypy checks.

Reply:
I added a more specific typing as @WardBrian suggested elsewhere.